nav2_waypoint_follower/src/waypoint_follower.cpp

```diff
@@ -0,0 +1,224 @@
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+WaypointFollower::on_error(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_FATAL(get_logger(), "Lifecycle node entered error state");
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+WaypointFollower::on_shutdown(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Shutting down");
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+void
+WaypointFollower::followWaypoints()
+{
+  auto goal = action_server_->get_current_goal();
+  auto feedback = std::make_shared<ActionT::Feedback>();
+  auto result = std::make_shared<ActionT::Result>();
+
+  // Check if request is valid
+  if (!action_server_ || !action_server_->is_server_active()) {
+    RCLCPP_DEBUG(get_logger(), "Action server inactive. Stopping.");
+    return;
+  }
+
+  rclcpp::Rate r(10);
+  uint goal_index = 0;
+  bool new_goal = true;
+
+  while (rclcpp::ok()) {
+    // Check if asked to stop processing action
+    if (action_server_->is_cancel_requested()) {
+      RCLCPP_INFO(get_logger(), "Cancelling action.");
+      action_server_->terminate_goals();
+      return;
+    }
+
+    // Check if asked to process another action
+    if (action_server_->is_preempt_requested()) {
+      RCLCPP_INFO(get_logger(), "Preempting the goal pose.");
+      goal = action_server_->accept_pending_goal();
+      goal_index = 0;
+      new_goal = true;
+    }
+
+    // Check if we need to send a new goal
+    if (new_goal) {
+      new_goal = false;
+      ClientT::Goal client_goal;
+      client_goal.pose = goal->poses[goal_index];
+
+      auto send_goal_options = rclcpp_action::Client<ClientT>::SendGoalOptions();
+      send_goal_options.result_callback =
+        std::bind(&WaypointFollower::resultCallback, this, std::placeholders::_1);
+      send_goal_options.goal_response_callback =
+        std::bind(&WaypointFollower::goalResponseCallback, this, std::placeholders::_1);
+      auto future_goal_handle =
+        nav_to_pose_client_->async_send_goal(client_goal, send_goal_options);
+      current_goal_status_ = ActionStatus::PROCESSING;
+    }
+
+    feedback->current_waypoint = goal_index;
+    action_server_->publish_feedback(feedback);
+
+    if (current_goal_status_ == ActionStatus::FAILED) {
+      if (stop_on_failure_) {
+        RCLCPP_WARN(get_logger(), "Failed to process waypoint %i in waypoint "
+          "list and stop on failure is enabled."
+          " Terminating action.", goal_index);
+        action_server_->terminate_goals();
+        return;
+      } else {
+        RCLCPP_INFO(get_logger(), "Failed to process waypoint %i,"
+          " moving to next.", goal_index);
+      }
+    } else if (current_goal_status_ == ActionStatus::SUCCEEDED) {
+      RCLCPP_INFO(get_logger(), "Succeeded processing waypoint %i, "
+        "moving to next.", goal_index);
+    } else if (current_goal_status_ != ActionStatus::PROCESSING) {
```

Comment:
We're moving to the next waypoint while the current one is still processing? Is that what this means?

Reply:
If theres any end condition (failed, succeeded) and we make it to this point (so stop on failure is off) we start processing the next waypoint

Reply:
The current goal status is the status of the nav to pose action. For this set of logic, the nav to pose action has already in some way terminated